src/main/java/org/truffleruby/core/array/ArrayBuilderNode.java

```diff
@@ -26,520 +31,236 @@
-        private final ConditionProfile otherDoubleStoreProfile = ConditionProfile.createBinaryProfile();
-
-        public DoubleArrayBuilderNode(int expectedLength) {
-            this.expectedLength = expectedLength;
-        }
-
-        @Override
-        public Object start() {
-            return new double[expectedLength];
-        }
+    @ImportStatic(ArrayGuards.class)
+    public abstract static class AppendArrayNode extends ArrayBuilderBaseNode {
 
-        @Override
-        public Object start(int length) {
-            if (length > expectedLength) {
-                CompilerDirectives.transferToInterpreterAndInvalidate();
-                return restart(length, length + " > " + expectedLength + " (expected)");
-            }
-
-            return new double[expectedLength];
-        }
-
-        @Override
-        public Object ensure(Object store, int length) {
-            if (length > ((double[]) store).length) {
-                CompilerDirectives.transferToInterpreterAndInvalidate();
-                return ensureFallback(store, length);
-            }
-            return store;
+        public static AppendArrayNode create() {
+            return AppendArrayNodeGen.create();
         }
 
-        @Override
-        public Object appendArray(Object store, int index, DynamicObject array) {
-            Object otherStore = Layouts.ARRAY.getStore(array);
+        public abstract Object executeAppend(Object array, int index, DynamicObject value);
 
-            if (otherStore == null) {
-                return store;
+        @Specialization(guards = { "strategy.matchesStore(array)", "strategy.matches(other)" }, limit = "1")
+        public Object appendCompatibleType(Object array, int index, DynamicObject other,
+                @Cached("ofStore(array)") ArrayStrategy strategy,
+                @Cached("createBinaryProfile()") ConditionProfile expandProfile) {
+            ArrayMirror mirror = strategy.newMirrorFromStore(array);
+            final ArrayMirror otherMirror = strategy.newMirror(other);
+            final int otherSize = Layouts.ARRAY.getSize(other);
+            final int neededSize = index + otherSize;
+            if (expandProfile.profile(neededSize > mirror.getLength())) {
+                replaceNodes(strategy, index);
+                final int capacity = Math.max(mirror.getLength(), neededSize);
+                mirror = mirror.copyArrayAndMirror(capacity);
             }
-
-            if (otherDoubleStoreProfile.profile(otherStore instanceof double[])) {
-                System.arraycopy(otherStore, 0, store, index, Layouts.ARRAY.getSize(array));
-                return store;
+            otherMirror.copyTo(mirror, 0, index, otherSize);
+            return mirror.getArray();
+        }
+
+        @Specialization(guards = { "arrayStrategy != otherStrategy", "arrayStrategy == generalized",
+                                   "arrayStrategy.matchesStore(array)", "otherStrategy.matches(other)" },
```

Comment:
In what can we have these conditions in the guards?
Seems like the Specialization above already handles this but its limit is too low (1).